### PR TITLE
Publish contact updates & contact code periodically

### PR DESCRIPTION
**Pull request by cammellos**
## Contact updates
Currently it's very easy for contact details to get out of sync, the
simplest example is:
A & B are contacts.
A changes name.
B receives the updated name.
B re-install the app.
Until A changes name again, B will not see their name, picture and won't
be able to send push notifications.

This PR changes the behavior to publish account informations to contacts
every 48 hrs, to add some redundancy in this cases.

## Contact code

This PR also publishes contact codes periodically to a public channel, once every 12hrs.
This allow other peers to fetch our information so that they can contact securely using PFS and include all our devices.
This change leaks some metadata, namely if you have been online in the past 12 hours, but it's the same as publishing on swarm/ipfs, we are trading off some darkness for stronger encryption and usability.

### Testing

It's a bit hard to test as the problem only shows after the contact request is not on the mailserver anymore.
Theoretically it goes like this:

1) Create account A
2) Create account B
3) Add B to contacts from A
4) You should now see contact details of A on B
5) Change clock on B to +2 days
6) Unistall/Reinstall  the app on B
7) Login, make sure you are connected, wait until messages are fetched on the mailserver. Check that A is not in your contacts (any message should not have their name). Make sure you can receive messages.
8) Change clock on A +2 days, make sure you are online and connected.
9) Wait a few minutes (+/- 5 mins).

You should see contact information of A on B eventually.

Mind that any time you change the clock, things might get a bit weird (you might go offline/online/offline etc). Generally login/logout fixes it.

status: ready

### CODE REVIEW

File: src/status_im/accounts/update/publisher.cljs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+(defn publish-update [web3 done-fn]
+  (let [db @re-frame.db/app-db
+        my-public-key (get-in db [:account/account :public-key])
+        peers-count (:peers-count db)
+        now (datetime/timestamp)
+        last-updated (get-in
+                      db
+                      [:account/account :last-updated])]
+    (when (and (pos? peers-count)
+               (not (nil? last-updated))
```
Comment (rasom) [inline src/status_im/accounts/update/publisher.cljs]: can we use here `number?` or `int?`?

File: src/status_im/transport/shh.cljs [hunk 2]
```diff
@@ -67,22 +67,24 @@
         (log/debug :shh/post-success))
       (re-frame/dispatch [error-event err resp]))))
 
+(defn send-direct-message! [web3 direct-message success-event error-event count]
```
Comment (rasom) [inline src/status_im/transport/shh.cljs:72]: `count` is not used

File: src/status_im/accounts/update/core.cljs [hunk 3]
```diff
@@ -1,9 +1,49 @@
 (ns status-im.accounts.update.core
-  (:require [status-im.data-store.accounts :as accounts-store]
+  (:require [re-frame.core :as re-frame]
+            [status-im.utils.async :as async-util]
+            [status-im.data-store.accounts :as accounts-store]
+            [status-im.utils.datetime :as datetime]
```
Comment (rasom) [inline src/status_im/accounts/update/core.cljs]: `datetime`, `re-frame`, `async-util` are not used

**Review by rasom — APPROVED** [review]
noop: 36m

File: src/status_im/accounts/update/publisher.cljs (hunk 1 above)
Comment (cammellos) [reply src/status_im/accounts/update/publisher.cljs]: Thanks, I changed it to `pos?` as I noticed sometimes is `0`, in such cases we don't want to publish a contact update as it means we never updated our info.

**Comment by cammellos** [comment]
@rasom I have updated to share the contact code, could you please review again?
Thanks!

File: src/status_im/accounts/update/publisher.cljs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            [status-im.pairing.core :as pairing]
+            [status-im.utils.async :as async-util]
+            [status-im.data-store.accounts :as accounts-store]
+            [status-im.utils.datetime :as datetime]
+            [status-im.transport.message.protocol :as protocol]
+            [status-im.transport.message.contact :as message.contact]
+            [status-im.transport.shh :as shh]
+            [status-im.utils.fx :as fx]))
+
+(defonce polling-executor (atom nil))
```
Comment (rasom) [inline src/status_im/accounts/update/publisher.cljs]: `polling-executor` is not used

File: src/status_im/accounts/update/publisher.cljs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  (:require [re-frame.core :as re-frame]
+            [re-frame.db]
+            [status-im.constants :as constants]
+            [status-im.accounts.update.core :as accounts]
+            [status-im.pairing.core :as pairing]
+            [status-im.utils.async :as async-util]
+            [status-im.data-store.accounts :as accounts-store]
+            [status-im.utils.datetime :as datetime]
+            [status-im.transport.message.protocol :as protocol]
+            [status-im.transport.message.contact :as message.contact]
```
Comment (rasom) [inline src/status_im/accounts/update/publisher.cljs]: `message.contact`, `protocol`,  `datetime`, `accounts-store`, `async-util`, `re-frame` are not used

File: src/status_im/pairing/core.cljs [hunk 6]
```diff
@@ -1,24 +1,70 @@
 (ns status-im.pairing.core
   (:require [re-frame.core :as re-frame]
+            [re-frame.db]
             [clojure.string :as string]
             [status-im.i18n :as i18n]
+            [taoensso.timbre :as log]
             [status-im.utils.fx :as fx]
             [status-im.ui.screens.navigation :as navigation]
+            [status-im.utils.async :as async-util]
+            [status-im.utils.datetime :as datetime]
```
Comment (rasom) [inline src/status_im/pairing/core.cljs]: `datetime`,`async-util` are not used

File: src/status_im/utils/publisher.cljs [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+            [re-frame.db]
+            [status-im.constants :as constants]
+            [status-im.accounts.update.publisher :as accounts]
+            [status-im.pairing.core :as pairing]
+            [status-im.utils.async :as async-util]
+            [status-im.data-store.accounts :as accounts-store]
+            [status-im.utils.datetime :as datetime]
+            [status-im.transport.message.protocol :as protocol]
+            [status-im.transport.message.contact :as message.contact]
+            [status-im.transport.shh :as shh]
```
Comment (rasom) [inline src/status_im/utils/publisher.cljs]: `shh`, `message.contact`, `protocol`, `accounts-store`, `constants` are not used

**Review by rasom — APPROVED** [review]
review time: 22m

File: src/status_im/accounts/update/core.cljs [hunk 8]
```diff
@@ -1,9 +1,46 @@
...
+  [{:keys [db] :as cofx} chat-id payload]
+  (when-let [chat (get-in cofx [:db :transport/chats chat-id])]
+    (let [updated-chat  (assoc chat :resend? "contact-update")
+          tx            [(transport-store/save-transport-tx {:chat-id chat-id
+                                                             :chat    updated-chat})]
+          success-event [:transport/contact-message-sent chat-id]]
+      (fx/merge cofx
+                {:db (assoc-in db
+                               [:transport/chats chat-id :resend?]
+                               "contact-update")
```
Comment (mandrigin) [inline src/status_im/accounts/update/core.cljs:24]: that does the same thing as `let [updated-chat ....` a few lines on top of it. Is there any way to join them?

**Review by mandrigin — APPROVED** [review]
overall looks good, small notes

File: src/status_im/accounts/update/core.cljs (hunk 8 above)
Comment (cammellos) [reply src/status_im/accounts/update/core.cljs:24]: I'll have a look, thanks!

**Comment by churik** [comment]
@cammellos 
tested scenario described in https://github.com/status-im/status-react/pull/7342#issue-247671292 - everything works as expected: 
1) User A (Mac OS), User B (Android 7.0), no profile picture is set for User A
2) User A (IOS 11.4), User B (Mac OS), custom profile picture is set for User A

As regression tested device pairing - everything looks fine from my side.

**Comment by cammellos** [comment]
@churik thanks, merged!

**Pull request merged**